File: opentsdb-proxy/src/main/java/net/opentsdb/proxy/modules/ProxyModule.java
package net.opentsdb.proxy.modules;

import com.google.inject.AbstractModule;
import com.google.inject.Injector;
import com.google.inject.Provides;
import com.google.inject.Singleton;
import com.google.inject.name.Named;
import com.google.inject.name.Names;
import net.opentsdb.proxy.clients.Client;
import net.opentsdb.proxy.netty.PipelineFactory;
import org.jboss.netty.channel.ChannelFactory;
import org.jboss.netty.channel.group.ChannelGroup;
import org.jboss.netty.channel.group.DefaultChannelGroup;
import org.jboss.netty.channel.socket.nio.NioServerSocketChannelFactory;

import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.util.Properties;
import java.util.concurrent.Executors;

/**
 * Guice module for binding instances together
 */
public class ProxyModule extends AbstractModule {
  private static final String PROXY_PORT = "tsdb.proxy.port";
  private static final String CLIENT_CLASS = "tsdb.proxy.client.class";

  private final Properties properties;

  public ProxyModule(Properties properties) {
    this.properties = properties;
  }

  @Override
  protected void configure() {
    //Magic! loads the properties as something that can be accessed via @Named
    Names.bindProperties(binder(), properties);
    bind(PipelineFactory.class).asEagerSingleton();
  }

  @Provides
  @Singleton
  Client provideClient(Injector injector, @Named(CLIENT_CLASS) Class<Client> clientClass) {
    return injector.getInstance(clientClass);
  }

  @Provides
  @Singleton
  Properties provideProperties() {
    return properties;
  }

  @Provides
  @Singleton
  SocketAddress provideSocketAddress(@Named(PROXY_PORT) Integer port) {
    return new InetSocketAddress(port);
  }

  @Provides
  @Singleton
  ChannelGroup provideChannelGroup() {
    return new DefaultChannelGroup("opentsdb-proxy");
  }

  @Provides
  @Singleton
  ChannelFactory provideChannelFactory() {
    return new NioServerSocketChannelFactory(Executors.newCachedThreadPool(), Executors.newCachedThreadPool());
  }
}

File: opentsdb-proxy/src/main/scala/net/opentsdb/proxy/modules/ProxyModule.scala
package net.opentsdb.proxy.modules

import java.util.Properties
import com.google.inject.{Injector, Singleton, Provides, AbstractModule}
import net.codingwell.scalaguice.ScalaModule
import net.opentsdb.proxy.clients.Client
import java.net.{InetSocketAddress, SocketAddress}
import org.jboss.netty.channel.group.{DefaultChannelGroup, ChannelGroup}
import org.jboss.netty.channel.ChannelFactory
import org.jboss.netty.channel.socket.nio.NioServerSocketChannelFactory
import java.util.concurrent.Executors
import net.opentsdb.proxy.netty.PipelineFactory
import com.google.inject.name.{Names, Named}

class ProxyModule(props: Properties) extends AbstractModule with ScalaModule {
  def configure() {
    Names.bindProperties(binder(), props)
    bind(classOf[PipelineFactory]).asEagerSingleton()
  }

  @Provides @Singleton def provideClient(injector: Injector, @Named("tsdb.proxy.client.class") clientClass: Class[Client]): Client = injector.getInstance(clientClass)

  @Provides @Singleton def provideProperties: Properties = props

  @Provides @Singleton def provideSocketAddress(@Named("tsdb.proxy.port") port: Int): SocketAddress = new InetSocketAddress(port)

  @Provides @Singleton def provideChannelGroup: ChannelGroup = new DefaultChannelGroup("opentsdb-proxy")

  @Provides @Singleton def provideChannelFactory: ChannelFactory = new NioServerSocketChannelFactory(Executors.newCachedThreadPool(), Executors.newCachedThreadPool())
}
